### ISSUE TITLE
spec should queue tasks to expose attribute changes on ServiceWorker and ServiceWorkerRegistration

### ISSUE THREAD

**Issue by wanderview** [slightlyoff/ServiceWorker#860]
The spec has a couple DOM objects that are exposed to both main thread window and workers such that they represent the same underlying object.  ServiceWorker and ServiceWorkerRegistration are the main ones I'm aware of right now.

The attribute getters for both of these objects are written like this:

```
installing attribute must return the result of running these steps or their equivalent:
    1) Let installingWorker be the ServiceWorker object that represents the service worker
        registration’s installing worker.
    2) Return installingWorker. 
```

Essentially as a synchronous getter on the underlying shared data.  There is no explicit concurrency protection here.

The most likely way for implementations to actual write the code, however, is to queue tasks to update the attributes on the exposed DOM objects.  We should update the spec to be explicit about this.

This is important because whether we queue a task or not can have observable effects based on when statechange and updatefound events fire.  Those events are executed through queued tasks.  Using queued tasks for the attribute changes would ensure a consistent view of shared data by content script.

**Comment by NekR** [comment]
> * Bug in Gecko: registration.installing and activating currently returns null

Hm.. always? I am using it here: https://github.com/NekR/offline-plugin/blob/v3/tpls/runtime-template.js#L15

Does it mean that this won't work in Firefox? 

I am trying to achive this, which works currently fine in Chrome:

```js
const runtime = require('offline-plugin/runtime');

runtime.install({
  onUpdating: () => {
    console.log('SW Event:', 'onUpdating');
  },
  onUpdateReady: () => {
    console.log('SW Event:', 'onUpdateReady');
    runtime.applyUpdate();
  },
  onUpdated: () => {
    console.log('SW Event:', 'onUpdated');
    window.location.reload();
  },

  onUninstalled: () => {
    console.log('SW Event:', 'onUninstalled');
  }
});
```

**Comment by NekR** [comment]
One thing that I noticed in Chrome is that after `updatefound` and having `registration.installing` SW, that `installing` SW may fire `statechange` to `installing`. This is why I have to use such guards as `ignoreInstalling` (state) after I already found installing Service Worker in registration.

I guess it's just a race, but a bad one.

**Comment by wanderview** [comment]
>>        Bug in Gecko: registration.installing and activating currently returns null

> Hm.. always? I am using it here: https://github.com/NekR/offline-plugin/blob/v3/tpls/runtime-template.js#L15

In gecko we expose registration.installing, etc on main thread windows, but not on worker threads yet.  So if you use `self.registration.active` in the service worker itself you will always get null in firefox at the moment.

Our bug to fix:

  https://bugzilla.mozilla.org/show_bug.cgi?id=1113522

**Comment by wanderview** [comment]
Gecko bug to queue tasks: https://bugzilla.mozilla.org/show_bug.cgi?id=1257977

**Comment by NekR** [comment]
> In gecko we expose registration.installing, etc on main thread windows, but not on worker threads yet. So if you use self.registration.active in the service worker itself you will always get null in firefox at the moment.

Thanks for info. I thought Gecko doesn't have it in main thread. No so much important in SW thread at a moment.

**Comment by jungkees** [comment]
Addressed: 3266af01ca3ab891d330a86ba3182bc5dae6c9b2
